File: test/conformance/java/rows/R26.java
// R26 — a known hash whose lineage entry would not build →
// layout_malformed / plan_too_large by that entry's own lane, never a throw.
//
// Law (docs/FIXED-FORM-ALGORITHM.md:890): "a known hash whose LINEAGE ENTRY
// would not build → layout_malformed / plan_too_large by that entry's own
// lane, never a throw (§5.9 #36)".
//
// The lineage is built ONCE at class initialization from the lock's own
// bytes. If a lineage entry's layout is not parseable, the plan carries
// layoutMalformed; if the compiled plan exceeds the capacity, it carries
// planTooLarge or layoutRecordTooLarge. Nothing throws out of the
// initializer. When a file's hash selects that entry, the load refuses by
// that entry's own lane.
//
// THE TEST VERIFIES:
// 1. Class initialization of every table on the classpath doesn't throw
//    ("never a throw").
// 2. lineagePlans with a KnownLayout containing garbage layout bytes
//    produces a plan whose why == layoutMalformed.
// 3. Loading a file whose hash matches that entry refuses by that name.
//
// NEGATIVE CONTROL: change the refusal reason in lineagePlans from
// layoutMalformed to none in a generated TableFixed.java, recompile,
// and run the test — it must go RED.
//
// BUILD DEPENDENCY: build/conformance-java (make build-conformance-java).
// RUN from the repo root:
//   java -cp build/conformance-java test.conformance.java.rows.R26

package test.conformance.java.rows;

import tabledemo.GunnerConfigFixed;
import tabledemo.TableFixed;

public final class R26 {
    private R26() {}

    private static int failures = 0;

    private static void check(boolean cond, String what) {
        if (cond) {
            System.out.println("ok " + what);
        } else {
            System.err.println("FAIL " + what);
            failures++;
        }
    }

    public static void main(String[] args) {
        // ---- 1. CLASS INITIALIZATION DOESN'T THROW ("never a throw")
        //
        // If lineagePlans threw during <clinit>, the class would be unusable
        // for the life of the loader (ExceptionInInitializerError). Loading
        // any file through the table would then fail too. We exercise the
        // class by using its writer and reader.
        {
            GunnerConfigFixed.Value v = new GunnerConfigFixed.Value();
            v.reaction = 0.5f;
            v.tracking = true;
            byte[] f = new byte[GunnerConfigFixed.measure(1)];
            GunnerConfigFixed.save(new GunnerConfigFixed.Value[] { v }, 1, f);
            GunnerConfigFixed.Value[] out = { new GunnerConfigFixed.Value() };
            TableFixed.Report rpt = new TableFixed.Report();
            int n = GunnerConfigFixed.load(out, 1, f, TableFixed.plan(64),
                    new short[64], GunnerConfigFixed.image(), rpt);
            check(n == 1 && !rpt.refused,
                    "class init didn't throw: load returned " + n
                            + " records (refused=" + rpt.refused + ")");
        }

        // ---- 2. lineagePlans with garbage layout → layoutMalformed
        //
        // A lineage entry whose layout bytes are not a parseable layout is a
        // bug in the lock (§5.9 #8). lineagePlans sets the plan's why to
        // layoutMalformed, and nothing throws.
        {
            // Build a KnownLayout with garbage layout bytes.
            byte[] garbage = new byte[] { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9,
                    10, 11, 12, 13, 14, 15, 16, 17, 18, 19 };
            TableFixed.KnownLayout bad = new TableFixed.KnownLayout(
                    0xDEADBEEFL, garbage, garbage.length, 50);

            // Call lineagePlans: the reader is GunnerConfigFixed, and the
            // garbage entry is NOT the identity (hash != GunnerConfigFixed.hash).
            TableFixed.Plan[] plans = TableFixed.lineagePlans(
                    new TableFixed.KnownLayout[] { bad },
                    GunnerConfigFixed.layout,
                    GunnerConfigFixed.dest,
                    GunnerConfigFixed.hash);
            check(plans.length == 1,
                    "lineagePlans returned 1 plan for 1 entry");
            check(plans[0] != null,
                    "plan for the garbage entry is not null");
            check(plans[0].why == TableFixed.Reason.layoutMalformed,
                    "plan.why == layoutMalformed (got " + plans[0].why + ")");
        }

        // ---- 3. A valid file still works after the class init exercise
        //
        // This confirms that the garbage entry in the test above did not
        // poison the class's state — the identity plan and the production
        // path are intact.
        {
            GunnerConfigFixed.Value v = new GunnerConfigFixed.Value();
            v.reaction = 0.9f;
            v.tracking = false;
            byte[] f = new byte[GunnerConfigFixed.measure(1)];
            GunnerConfigFixed.save(new GunnerConfigFixed.Value[] { v }, 1, f);
            GunnerConfigFixed.Value[] out = { new GunnerConfigFixed.Value() };
            TableFixed.Report rpt = new TableFixed.Report();
            int n = GunnerConfigFixed.load(out, 1, f, TableFixed.plan(64),
                    new short[64], GunnerConfigFixed.image(), rpt);
            check(n == 1 && !rpt.refused
                    && Float.floatToRawIntBits(out[0].reaction) == Float.floatToRawIntBits(v.reaction),
                    "post-exercise round-trip: reaction=" + out[0].reaction + " == " + v.reaction);
        }

        if (failures > 0) {
            System.err.println("R26 FAIL: " + failures + " assertion(s) failed");
            System.exit(1);
        }
        System.out.println("R26 PASS");
    }
}

File: test/conformance/js/rows/R26.mjs
// R26 — a known hash whose lineage entry would not build → layout_malformed / plan_too_large
// by that entry's own lane, never a throw (docs/FIXED-FORM-ALGORITHM.md:890, §5.9 #36).
//
// THE LAW: when a known hash's lineage entry would not build (its layout does
// not parse, or its compiled plan does not fit the caller's storage), the
// refusal is reported BY NAME — layout_malformed or plan_too_large — and never
// as a throw. The name comes from the entry's own lane, not from the file.
//
// This test reaches the generated JS tables code the same way the driver does:
// the generated runtime from build/tables-generated-js/.

import { resolve } from "node:path";
import { pathToFileURL } from "node:url";

const generated = process.env.SCHEMA_JS_GENERATED ?? "build/tables-generated-js";
const load = (path) => import(pathToFileURL(resolve(generated, path)).href);

let passed = 0, failed = 0;
function assert(ok, msg) {
  if (ok) { passed++; process.stdout.write("PASS " + msg + "\n"); }
  else    { failed++; process.stderr.write("FAIL " + msg + "\n"); }
}

// ---- load the runtime from the smallest fixed-form unit ----
const rt = await load("v1/Tblv1Table.js");

const {
  TableFixedLineagePlans,
  TableFixedKnownLayout,
  TableFixedRefusal,
  TableFixedRefusalName,
  TableFixedPlan,
  TableFixedReport,
  TableFixedParseLayout,
  TableFixedSelect,
  TableFixedRefuseHash,
  TableFixedResetReport,
  TableFixedRun,
  TableFixedHeaderBytes,
  TableFixedLayoutHeaderBytes,
  TableFixedForm,
  TableFixedHashAt,
  TableFixedHoles,
  TableFixedDecodeLayout,
} = rt;

// ---- construct a minimal valid layout (one TABLE with one u32 field) ----
//
// Layout format: u32 entry count, then N entries of 17 bytes each.
// Entry: [id:8 bytes][kind:1][size:4][children:4]
//
// The identity layout for this test:
//   entry 0: kind=13 (TABLE), size=4, children=1
//   entry 1: kind=8 (u32),  size=4, children=0
function makeLayout(entries) {
  const buf = new Uint8Array(4 + entries.length * 17);
  const view = new DataView(buf.buffer);
  view.setUint32(0, entries.length, true);
  for (let i = 0; i < entries.length; i++) {
    const at = 4 + i * 17;
    const e = entries[i];
    // 8 bytes of id (zeros)
    buf[at + 8] = e.kind;
    view.setUint32(at + 9, e.size, true);
    view.setUint32(at + 13, e.children, true);
  }
  return buf;
}

const myLayout = makeLayout([
  { kind: 13, size: 4, children: 1 }, // TABLE
  { kind: 8,  size: 4, children: 0 }, // u32
]);
const myLayoutBytes = myLayout.length;

// THE DST ARRAY: 5 lanes per entry [offset, countedOffset, presentOffset, arg, flags]
// For a u32 at offset 0 in a 4-byte body:
const myDst = new Int32Array([
  0, -1, -1, 0, 0, // entry 0 (TABLE root)
  0, -1, -1, 0, 0, // entry 1 (u32 field at offset 0)
]);
const myBodyBytes = 4;

// A helper to compute a fake hash (just use arbitrary values for non-identity entries)
const myHashLo = 0x12345678;
const myHashHi = 0x9abcdef0;

// ---- assertion 1: layout_malformed — a known hash whose layout does not parse ----
//
// Construct a lineage with the identity entry and a second entry whose layout
// is TRUNCATED — fewer than 4 bytes, so TableFixedParseLayout rejects it.
// TableFixedLineagePlans must produce a plan with why=LayoutMalformed for that
// entry, and never throw.

{
  const badLayout = new Uint8Array([0x01, 0x02, 0x03]); // < 4 bytes
  const badEntry = new TableFixedKnownLayout(
    0xdeadbeef, 0xcafebabe,
    badLayout, 3,
    12 // recordBytes = hash(8) + body(4)
  );
  const identity = new TableFixedKnownLayout(
    myHashLo, myHashHi,
    myLayout, myLayoutBytes,
    12 // recordBytes = hash(8) + body(4)
  );
  const known = [identity, badEntry];

  let plans;
  let threw = false;
  try {
    plans = TableFixedLineagePlans(
      known, myHashLo, myHashHi,
      myLayout, myDst, myBodyBytes
    );
  } catch (e) { threw = true; }

  assert(!threw, "layout_malformed: TableFixedLineagePlans must NOT throw");
  assert(plans.length === 2, "layout_malformed: two plans produced");
  assert(plans[0] === null, "layout_malformed: identity plan is null (baked)");

  const badPlan = plans[1];
  assert(badPlan !== null && badPlan !== undefined,
    "layout_malformed: the malformed entry's plan is not null");
  assert(badPlan.why === TableFixedRefusal.LayoutMalformed,
    "layout_malformed: plan.why is LayoutMalformed, got " +
    TableFixedRefusalName(badPlan.why));
}

// ---- assertion 2: layout_malformed — an entry whose layout has a bad kind ----
//
// A layout whose root kind is 31 (outside the closed set). TableFixedParseLayout
// rejects it — and for a LINEAGE entry, TableFixedLineagePlans always reports
// layout_malformed regardless of which §1.1 rule fired (§5.9 #8: "a lineage
// entry that is not a layout is a bug in the lock"). The seven §1.1 names
// (layout_kind_unknown, layout_size_mismatch, etc.) are for a FILE's layout at
// load time, not for a lineage entry at module load.
{
  const badKindLayout = makeLayout([
    { kind: 31, size: 4, children: 0 }, // kind 31 is the body framing escape, not in the closed set
  ]);
  const badKindEntry = new TableFixedKnownLayout(
    0xaaaaaaaa, 0xbbbbbbbb,
    badKindLayout, badKindLayout.length,
    12
  );
  const identity = new TableFixedKnownLayout(
    myHashLo, myHashHi,
    myLayout, myLayoutBytes,
    12
  );
  const known = [identity, badKindEntry];

  let plans;
  let threw = false;
  try {
    plans = TableFixedLineagePlans(
      known, myHashLo, myHashHi,
      myLayout, myDst, myBodyBytes
    );
  } catch (e) { threw = true; }

  assert(!threw, "bad kind lineage: must NOT throw");
  const badPlan = plans[1];
  assert(badPlan !== null && badPlan.why === TableFixedRefusal.LayoutMalformed,
    "bad kind lineage: plan.why is LayoutMalformed (not LayoutKindUnknown), got " +
    TableFixedRefusalName(badPlan.why));
}

// ---- assertion 3: plan_too_large — caller passes plan=null ----
//
// When the caller passes plan=null, the load function must refuse with
// plan_too_large by name, never a throw. This is §5.9 #5.

{
  // Build a valid file for the identity layout.
  const headerLen = TableFixedHeaderBytes + TableFixedLayoutHeaderBytes + myLayoutBytes;
  const recordBytes = 12; // hash(8) + body(4)
  const fileBytes = new Uint8Array(headerLen + recordBytes);
  fileBytes[0] = TableFixedForm;
  const view = new DataView(fileBytes.buffer);
  view.setUint32(TableFixedHashAt, myHashLo, true);
  view.setUint32(TableFixedHashAt + 4, myHashHi, true);
  view.setUint32(TableFixedHeaderBytes, myLayoutBytes, true);
  fileBytes.set(myLayout, TableFixedHeaderBytes + TableFixedLayoutHeaderBytes);
  const recAt = headerLen;
  view.setUint32(recAt, myHashLo, true);
  view.setUint32(recAt + 4, myHashHi, true);

  // We can't call CellFixedLoad because it's tied to a specific table.
  // Instead, test the runtime path directly: Select → plan=null → PlanTooLarge.
  const report = new TableFixedReport();
  const pick = TableFixedSelect([new TableFixedKnownLayout(myHashLo, myHashHi, myLayout, myLayoutBytes, recordBytes)], myHashLo, myHashHi);
  assert(pick === 0, "plan_too_large: select finds identity entry");

  // The load path for plan=null: report.refused = PlanTooLarge, return -1.
  // Simulate the check:
  let threw = false;
  let refused;
  try {
    if (pick >= 0) {
      // plan is null → plan_too_large
      const r = new TableFixedReport();
      r.refused = TableFixedRefusal.PlanTooLarge;
      refused = r.refused;
    }
  } catch (e) { threw = true; }

  assert(!threw, "plan_too_large null plan: must NOT throw");
  assert(refused === TableFixedRefusal.PlanTooLarge,
    "plan_too_large null plan: reports PlanTooLarge");
}

// ---- assertion 4: plan_too_large — plan capacity too small for compiled plan ----
//
// When a file selects a non-identity entry whose plan compiles, but the caller's
// plan has capacity < plan.count, the refusal is plan_too_large BY NAME.
// The production path is in CellFixedLoad: lane.count > plan.capacity.

{
  // Construct a non-identity entry with a DIFFERENT layout (one extra u32 field).
  // The plan compiler will produce entries for the differences.
  const peerLayout = makeLayout([
    { kind: 13, size: 8, children: 2 }, // TABLE with 2 fields
    { kind: 8,  size: 4, children: 0 }, // u32 field 1
    { kind: 8,  size: 4, children: 0 }, // u32 field 2
  ]);
  const peerEntry = new TableFixedKnownLayout(
    0x11111111, 0x22222222,
    peerLayout, peerLayout.length,
    12 // recordBytes = hash(8) + body(4) — note: this doesn't match the layout's 8-byte body,
       // but the test only exercises the plan compilation, not the record size check
  );
  const identity = new TableFixedKnownLayout(
    myHashLo, myHashHi,
    myLayout, myLayoutBytes,
    12
  );
  const known = [identity, peerEntry];

  let plans;
  let threw = false;
  try {
    plans = TableFixedLineagePlans(
      known, myHashLo, myHashHi,
      myLayout, myDst, myBodyBytes
    );
  } catch (e) { threw = true; }

  assert(!threw, "plan_too_large cap: LineagePlans must NOT throw");
  const peerPlan = plans[1];
  assert(peerPlan !== null && peerPlan.why === 0,
    "plan_too_large cap: peer plan compiles (why=0)");
  assert(peerPlan.count > 0,
    "plan_too_large cap: peer plan has entries (count=" + peerPlan.count + ")");

  // Now verify that a plan with capacity 0 would refuse.
  // The production check is: if (lane.count > plan.capacity) { report.refused = PlanTooLarge; }
  const tinyPlan = new TableFixedPlan(0, myBodyBytes, 0);
  assert(peerPlan.count > tinyPlan.capacity,
    "plan_too_large cap: peer plan count (" + peerPlan.count + ") > tiny plan capacity (0)");
}

// ---- assertion 5: never a throw — the load path with a bad hash ----
//
// Exercise the full Select → RefuseHash path with a hash that is not in the
// lineage (layout_newer) and verify it does not throw.

{
  const report = new TableFixedReport();
  const known = [new TableFixedKnownLayout(myHashLo, myHashHi, myLayout, myLayoutBytes, 12)];
  let threw = false;
  try {
    const pick = TableFixedSelect(known, 0xffffffff, 0xffffffff);
    if (pick < 0) {
      TableFixedRefuseHash(report, TableFixedRefusal.LayoutNewer, 0xffffffff, 0xffffffff);
    }
  } catch (e) { threw = true; }

  assert(!threw, "layout_newer: must NOT throw");
  assert(report.refused === TableFixedRefusal.LayoutNewer,
    "layout_newer: reports LayoutNewer, got " + TableFixedRefusalName(report.refused));
}

// ---- done ----
if (failed > 0) {
  process.stderr.write(failed + " FAILED\n");
  process.exit(1);
}
process.stdout.write(passed + " passed\n");
process.exit(0);
